File: solutions/0105_Construct_Binary_Tree_from_Preorder_and_Inorder_Traversal/recur.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> TreeNode:
        self.buildTree(preorder, inorder)

    def buildTree(self, preorder, inorder):
        if not preorder:
            return None
        root_val = preorder[0]
        root_index = inorder.index(root_val)
        root = TreeNode(root_val)
        num_left = root_index
        num_right = len(inorder) - 1 - root_index
        l_pre = preorder[1:num_left+1]
        r_pre = preorder[num_left+1:]
        l_in = inorder[:num_left]
        r_in = inorder[num_left+1:]
        root.left = self.buildTree(l_pre, l_in)
        root.right = self.buildTree(r_pre, r_in)
        return root

File: solutions/0105_Construct_Binary_Tree_from_Preorder_and_Inorder_Traversal/recur.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode buildTree(int[] preorder, int[] inorder) {
        return build(preorder, 0, preorder.length - 1,
                     inorder, 0, inorder.length - 1);
    }

    // preoder looks like root...left_childs...right_childs
    // inorder looks like ...left_childs...root...right_childs
    private TreeNode build(int[] preorder, int a, int b,
                           int[] inorder, int c, int d) {
        if (preorder == null || preorder.length == 0 || a > b || c > d) {
            return null;
        }
        TreeNode root = new TreeNode(preorder[a]);
        if (a == b || c == d) return root;
        int rootIndex = indexOf(inorder, preorder[a], c, d);
        int lenOfLeft = rootIndex - c;
        // then split preorder into two parts
        // left part
        root.left = build(preorder, a + 1, a + 1 + lenOfLeft - 1,
                          inorder,  c,     rootIndex - 1);
        root.right = build(preorder, a + 1 + lenOfLeft, b,
                           inorder,  rootIndex + 1,     d);
        return root;
    }

    private int indexOf(int[] arr, int value, int a, int b) {
        for (int i = a; i <= b; ++i) {
            if (arr[i] == value) return i;
        }
        return -1;
    }
}
